src/Parser/TypeParser.php

```diff
@@ -323,24 +323,54 @@ public function parseGeneric(TokenIterator $tokens, Ast\Type\IdentifierTypeNode
 	{
 		$tokens->consumeTokenType(Lexer::TOKEN_OPEN_ANGLE_BRACKET);
 		$tokens->tryConsumeTokenType(Lexer::TOKEN_PHPDOC_EOL);
-		$genericTypes = [$this->parse($tokens)];
+
+		$genericTypes = [];
+		$variances = [];
+
+		[$genericTypes[], $variances[]] = $this->parseGenericTypeArgument($tokens);
 
 		$tokens->tryConsumeTokenType(Lexer::TOKEN_PHPDOC_EOL);
 
 		while ($tokens->tryConsumeTokenType(Lexer::TOKEN_COMMA)) {
 			$tokens->tryConsumeTokenType(Lexer::TOKEN_PHPDOC_EOL);
 			if ($tokens->tryConsumeTokenType(Lexer::TOKEN_CLOSE_ANGLE_BRACKET)) {
 				// trailing comma case
-				return new Ast\Type\GenericTypeNode($baseType, $genericTypes);
+				return new Ast\Type\GenericTypeNode($baseType, $genericTypes, $variances);
 			}
-			$genericTypes[] = $this->parse($tokens);
+			[$genericTypes[], $variances[]] = $this->parseGenericTypeArgument($tokens);
 			$tokens->tryConsumeTokenType(Lexer::TOKEN_PHPDOC_EOL);
 		}
 
 		$tokens->tryConsumeTokenType(Lexer::TOKEN_PHPDOC_EOL);
 		$tokens->consumeTokenType(Lexer::TOKEN_CLOSE_ANGLE_BRACKET);
 
-		return new Ast\Type\GenericTypeNode($baseType, $genericTypes);
+		return new Ast\Type\GenericTypeNode($baseType, $genericTypes, $variances);
+	}
+
+
+	/**
+	 * @phpstan-impure
+	 * @return array{Ast\Type\TypeNode, Ast\Type\GenericTypeNode::VARIANCE_*}
+	 */
+	public function parseGenericTypeArgument(TokenIterator $tokens): array
+	{
+		if ($tokens->tryConsumeTokenType(Lexer::TOKEN_WILDCARD)) {
+			return [
+				new Ast\Type\IdentifierTypeNode('mixed'),
+				Ast\Type\GenericTypeNode::VARIANCE_BIVARIANT,
+			];
+		}
+
+		if ($tokens->tryConsumeTokenValue('contravariant')) {
+			$variance = Ast\Type\GenericTypeNode::VARIANCE_CONTRAVARIANT;
+		} elseif ($tokens->tryConsumeTokenValue('covariant')) {
+			$variance = Ast\Type\GenericTypeNode::VARIANCE_COVARIANT;
+		} else {
+			$variance = Ast\Type\GenericTypeNode::VARIANCE_INVARIANT;
```

Comment:
Some unsupported covariance name like `bullshitriant` isn't gonna cause a parse error and will silently fall back to invariant, right? Maybe we should address that or at least test it.

Reply:
It falls back to invariant and proceeds to parse the value as a type. Thus, 

- `Foo<bullshitriant>` creates a generic type with `IdentifierTypeNode('bullshitriant')` in it (supposedly failing later due to the unknown class). That seems correct to me, I don't think there's a reliable way to distinguish between a typo and a type on the parser level.
- `Foo<bullshitriant Type>` throws because 'bullshitriant' is parsed as a type and the parser expects a comma or a closing angle bracket to follow. I'll add this case to the test.